Refactor Code. Fix submit and shift buttons

TelhaBilal opened the pull request:
![gradio-keyboard-demo](https://github.com/user-attachments/assets/0a7d7db4-f177-417f-bb3d-4d23a7d11a28)

TelhaBilal committed:
Refactor + Make submit and shift working

TelhaBilal committed:
Remove Urdu from supported languages

Urdu is not yet made in refactored code

TelhaBilal committed:
Bound Key updates to key mouse down

mdsrqbl commented on frontend/keyboard.ts:
```diff
@@ -0,0 +1,49 @@
+        return this.langToKeysetMap[this.selectedLanguage].getKeys();
+    }
+
+    isCurLanguageRTL() {
```
`Cur` wtDzCurMean?

TelhaBilal replied:
Current. Active Language: since the keyboard supports multiple languages

mdsrqbl commented on frontend/ikeyboard.ts:
```diff
@@ -0,0 +1,16 @@
+import { Key } from "./keyset";
+
+export interface IKeyboard {
```
what does the `I` in `IKeyboard` mean?

TelhaBilal replied:
....Interface...

mdsrqbl commented on frontend/keyboard.ts:
```diff
@@ -0,0 +1,49 @@
+    container: boolean = true;
+    expanded: boolean = true;
+
+    private langToKeysetMap: object = {};
```
could you spell out `lang` and does including the word "Map" in a variable name follow the conventions?

TelhaBilal replied:
spell out lang: sure

Private member. Added "Map'" for better understandability that this member is supposed to be used as a mapping and I don't think it breaks any conventions, but I can remove the "Map" if you think it better

mdsrqbl commented on frontend/keyboard.ts:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    toggle(): void {
+        console.log(this);
```
Could you move the logging to dev/debug mode and disable it in the production mode?

TelhaBilal replied:
My bad.

mdsrqbl commented on frontend/keyset.ts:
```diff
@@ -0,0 +1,53 @@
+
+
+class Key {
+    upperStr: string = "";
+    middleStr: string = "";
+    lowerStr: string = "";
+    primaryStrIdx: number = 0;
```
Could you justify using data type in variable name?

TelhaBilal replied:
Open to ideas for what they should be called.

mdsrqbl commented on frontend/Index.svelte:
```diff
@@ -24,243 +26,119 @@
+      dir={keyboard.isCurLanguageRTL() ? "rtl" : "ltr"}
     />
-    <p class="transliteration" bind:this={transliteration}></p>
+    <p class="transliteration" bind:this={keyboard.transliteration}></p>
```
please add a label "Transliteration:" behind this text.

Note: Transliteration will always be in latin script (so LTR format only)

mdsrqbl commented on frontend/Index.svelte:
```diff
@@ -24,243 +26,119 @@
+  {min_width}
+  allow_overflow={false}
+  padding={true}
+>
```
which code formatter are you using?

mdsrqbl commented on frontend/english-keyset.ts:
```diff
@@ -0,0 +1,339 @@
+    'x': { code: "KeyX", keyCode: 88 },
+    'y': { code: "KeyY", keyCode: 89 },
+    'z': { code: "KeyZ", keyCode: 90 }
+};
```
Amazing work 👏

mdsrqbl commented on frontend/Index.svelte:
```diff
@@ -24,243 +26,119 @@
                 : 'inactive-key'}"
             >
-              {!isSpecialKey(key) ? languageKeyMap[currentLanguage]["shift"][key.toLowerCase()] || "\u200c" : "\u200c"}
+              {key.upperStr == "" ? "\u200c" : key.upperStr}
```
use of an invisible character `\u200c` seems quite "hacky". I wouldn't want to waste any time on perfecting this but see if you can figure out a better approach.

mdsrqbl commented on frontend/Index.svelte:
```diff
@@ -24,243 +26,119 @@
+      on:input={handleKeysChange}
+    >
       {#each keys as row, rowIndex}
+        <!-- {#each keys as row, rowIndex} -->
```
remove the commented code

mdsrqbl commented on frontend/Index.svelte:
```diff
@@ -24,243 +26,119 @@
 
   <div class="dropdown">
+    <!-- <button on:click={keyboard.toggle.bind(keyboard)} class="expand-button"> -->
     <button on:click={toggle} class="expand-button">
```
what's going on here?

mdsrqbl commented on frontend/Index.svelte:
```diff
@@ -24,243 +26,119 @@
       class="scroll-hide"
       bind:value
-      bind:this={textAreaEl}
+      bind:this={keyboard.targetElement}
```
is it possible to expose this element on the python side so the user can add the keyboard to a editable text area of their choice? (we can leave that for future versions.)

TelhaBilal replied:
Let's do it in another PR

mdsrqbl commented on frontend/Index.svelte:
```diff
@@ -24,243 +26,119 @@
+    <div
+      class="keyboard"
+      transition:fly={{ y: 20, duration: 200 }}
+      on:input={handleKeysChange}
```
what does the `handleKeysChange` function do?

TelhaBilal replied:
Updates the Keys(incase when one presses shift or caps or when the languages will change.)

mdsrqbl commented on frontend/english-keyset.ts:
```diff
@@ -0,0 +1,339 @@
+        curValue = curValue.slice(0, -1);
+    } else if (key_str === "Enter") {
+        curValue += "\n";
+    }
```
wouldn't all this logic get repeated for other languages as well? (except for upper-case/lower-case of course)
can you come up with a way to generalize it?

mdsrqbl commented on frontend/english-keyset.ts:
```diff
@@ -0,0 +1,339 @@
+    'z': { code: "KeyZ", keyCode: 90 }
+};
+
+function getModifiedValueOnKeyPress(curValue: string, key_str: string, states: object): string {
```
the value is not modified in all cases so is this the best function name?

TelhaBilal replied:
The idea is, on a particular keypress, if a value is not modified then the actual value is the modified value

mdsrqbl commented on frontend/english-keyset.ts:
```diff
@@ -0,0 +1,339 @@
+            return this.unShiftedKeys;
+        }
+    }
+}
```
add empty line at the end of file

mdsrqbl commented on frontend/ikeyboard.ts:
```diff
@@ -0,0 +1,16 @@
+    getKeys(): Key[][];
+    isCurLanguageRTL(): boolean;
+    getAllLanguageNames(): string[];
+}
```
add empty line at the end of file

mdsrqbl commented on frontend/keyset.ts:
```diff
@@ -0,0 +1,53 @@
+    }
+}
+
+export { Key, KeySet };
```
add empty line at the end of file

mdsrqbl commented on frontend/keyboard.ts:
```diff
@@ -0,0 +1,49 @@
+    getAllLanguageNames(): string[] {
+        return Object.keys(this.langToKeysetMap);
+    }
+}
```
add empty line at the end of file

mdsrqbl commented on frontend/english-keyset.ts:
```diff
@@ -0,0 +1,339 @@
+                    : key === "Tab" || key === "\\"
+                        ? 3
+                        : 2;
+    }
```
is this the best approach to define key spans? do you think a string to number object would be better?

TelhaBilal replied:
Maybe make a text to span mapping object?

mdsrqbl commented on frontend/english-keyset.ts:
```diff
@@ -0,0 +1,339 @@
+import { Key, KeySet } from "./keyset";
+
+
+export const charToCodesMapping = {
```
Should we add language name to this variable's name to avoid conflict when multiple keysets are imported?

TelhaBilal replied:
every character has its own code so even if we keep one map for all languages, it should be fine.

mdsrqbl commented on frontend/english-keyset.ts:
```diff
@@ -0,0 +1,339 @@
+        ["Caps", "A", "S", "D", "F", "G", "H", "J", "K", "L", ":", '"', "Enter"],
+        ["Shift", "Z", "X", "C", "V", "B", "N", "M", "<", ">", "?", "Shift"],
+        ["Ctrl", "Alt", "Space", "Alt", "Ctrl"],
+    ];
```
How do we ensure the key layout across multiple languages is consistent?

TelhaBilal replied:
This layout is just for a "simple english keyboard"
users are free to make their own keyboards with more or less keys or with different layouts

mdsrqbl commented on frontend/Index.svelte:
```diff
@@ -24,243 +26,119 @@
   export let scale: number | null = null;
   export let min_width: number | undefined = undefined;
   export let loading_status: LoadingStatus | undefined = undefined;
-  export let value_is_output = false;
   export let interactive: boolean;
-  export let languages: string[] = ["english", "urdu", "hindi"];
```
These exports are for the python interface. It may break these are removed.

mdsrqbl commented on frontend/Index.svelte:
```diff
@@ -24,243 +26,119 @@
-              class="upper {(isShiftDown || isCapsLockOn) && currentLanguage != 'english'
+              class="upper {key.primaryStrIdx == 0
                 ? 'active-key'
                 : 'inactive-key'}"
```
i think we should rename these classes because active usually refers to a pressed button.
But here we are only specifying if the character is prominent or greyed out.

mdsrqbl commented on frontend/english-keyset.ts:
```diff
@@ -0,0 +1,339 @@
+    }
+
+    private getKeyDownFunction(key_str: string) {
+        return (e: KeyboardEvent) => { console.log(key_str + " Key Down") };
```
is the logging added because the functionality isn't implemented yet (mention that in a comment above) or just for testing (disable it in production)?

TelhaBilal replied:
because functionality isn't implemented

mdsrqbl commented on frontend/english-keyset.ts:
```diff
@@ -0,0 +1,339 @@
+        return (e: MouseEvent) => { };
+    }
+
+    private getKeySpan(key: string): number {
+        return key === "Space"
+            ? 14
+            : key === "Shift"
+                ? 5
+                : key === "Delete" ||
+                    key === "Enter" ||
+                    key === "Ctrl" ||
+                    key === "Alt" ||
+                    key === "Caps"
+                    ? 4
+                    : key === "Tab" || key === "\\"
+                        ? 3
+                        : 2;
+    }
+
+    private isSpecialKey(key: string): boolean {
+        return [
+            "Delete",
+            "Enter",
+            "Shift",
+            "Caps",
+            "Tab",
+            "Ctrl",
+            "Alt",
+            "Win",
+            "Space",
+        ].includes(key);
+    }
```
wouldn't all this code get repeated for other languages as well?
can you come up with a way to generalize it?

TelhaBilal replied:
Depending on the other language keyboards made, a lot of it MIGHT be repeated, so generalizing it between other languages was the target in next iteration.

But there's also the case that the keysets can change and multiple types of keyboards for the same languages can be made.

For example,

There can be a keyboard which resembles the MAC Keyboard, with the "Command" key and everything.
There can be a keyboard without the capsLock key(like we have in smartphones)

So if we're looking to generalize this functionality, maybe we can make a "SimpleQWERTYKeyset" base class or something and move this code there, and then use that base class to make our "simple" keyboards, but I wont recommend forcing these keys on the entire Keyboard like the previous iteration was.

mdsrqbl commented on frontend/english-keyset.ts:
```diff
@@ -0,0 +1,339 @@
+            }
+        };
+
+        return () => {
+            customMouseDownFn();
+        };
```
interesting

TelhaBilal replied:
Ah,

It's like this because initially the anonymous function was accepting the mouse event param incase the customFunction needed it, but I removed it later. I'll change it

mdsrqbl commented on frontend/english-keyset.ts:
```diff
@@ -0,0 +1,339 @@
+
+        let customMouseDownFn = () => {
+
+            if (key_str.toLowerCase() in ["shift", "ctrl", "alt"]) {
```
could you capture <kbd>Caps Lock</kbd> as well?

TelhaBilal replied:
Caps lock is not in the scope of this PR.

I'll capture it in the next one.

mdsrqbl commented on frontend/Index.svelte:
```diff
@@ -24,243 +26,119 @@
+    <select
+      id="language"
+      class="language-selector"
+      bind:value={keyboard.selectedLanguage}
```
lets expose the selected language to python as well so the user can link the selected language with other elements or update the keyboard language based on the other element.

mdsrqbl commented on frontend/english-keyset.ts:
```diff
@@ -0,0 +1,339 @@
+                    key === "Alt" ||
+                    key === "Caps"
+                    ? 4
+                    : key === "Tab" || key === "\\"
```
add `|| key === "|"` for the shifted layout

mdsrqbl reviewed (changes requested):
Great work overall. Here are some further improvements:

- The <kbd>Shift</kbd> key currently works like how I imagined the <kbd>Caps Lock</kbd> key (stays active until pressed again). Generally, I've observed the <kbd>Shift</kbd> to go inactive as soon as any other key is pressed.
- <kbd>0</kbd> <kbd>1</kbd> <kbd>2</kbd> are not working.
- clicking <kbd>Enter</kbd> should also submit.

mdsrqbl commented:
- <kbd>0</kbd> <kbd>1</kbd> <kbd>2</kbd> are not working.
- clicking <kbd>Enter</kbd> should also submit.

TelhaBilal committed:
Re-arranged keyset and en-keyset modules. Turned Keyset into Interface

TelhaBilal committed:
Move to-be-shared functions between different languages to utils

TelhaBilal committed:
Formatting. Fix imports. Add caps state. Add undo shift

TelhaBilal committed:
Add caps lock functionality

TelhaBilal committed:
Fix span for pipe symbol

TelhaBilal committed:
rename langToKeysetMap to languagetoKeysetMap

TelhaBilal committed:
change <level>Str variable names to <level>Text

TelhaBilal committed:
Add the "onEnter: Function" param to keyboard

Param takes a function which is executed on "enter" press

If not provided(Defaults to null), Enter adds a new line

TelhaBilal commented:
Changes made:

- Renamed suggested variables
- Removed console.log
- Re-arranged keyset and english keyset code
- Fixed shift and made Caps lock working
- Moved to-be-shared functionality between languages to functions in utils
- Added an onEnter functionality to Keyboard so it can be used to submit and add a new line and is not restricted to gradio submissions only.
- Fix span of pipe symbol

TelhaBilal committed:
Add transliteration and transliteration label

TelhaBilal committed:
Make a text to span object for english language

TelhaBilal committed:
Add support for languages passed from python

TelhaBilal committed:
Fix formatting

mdsrqbl commented on frontend/keysets/utils.ts:
```diff
@@ -0,0 +1,99 @@
+import { charToCodesMapping } from "../char-codes";
+import { KeySet } from "./keyset";
+
+const simpleEnTextToSpans: Object = {
```
`Simple`? `En`?

TelhaBilal replied:
Simple: With basic qwerty layout and no fancy keys
En: English

mdsrqbl committed:
update author & readme to-do-s
